test/common/testlib.py

```diff
@@ -490,14 +490,37 @@ def intercept(test, err):
             if stopTestRun is not None:
                 stopTestRun()
 
-    def setUp(self, macaddr=None, memory_mb=None, cpus=None):
+    # additional_machines is an array of dictionaries, one for each additional machine to be created, e.g.:
+    # additional_machines = [ { machine: { 'image': 'openshift' }, 'start': { 'memory_mb': 1024 } } ]
+    def setUp(self, macaddr=None, memory_mb=None, cpus=None, additional_machines=None):
         self.machines = [ ]
         self.machine = self.new_machine()
-        self.machine.start(macaddr=macaddr, memory_mb=memory_mb, cpus=cpus)
+
+        self.machine.start(macaddr=macaddr, memory_mb=memory_mb, cpus=cpus, wait_for_ip=False)
+
+        if additional_machines:
+            # first create all additional machines, wait for them later
+            for machine_options in additional_machines:
+                if not 'machine' in machine_options:
+                    machine_options['machine'] = { }
+                if not 'start' in machine_options:
+                    machine_options['start'] = { }
+                machine = self.new_machine(**machine_options['machine'])
+                options = machine_options['start']
+                options['wait_for_ip'] = False
+                machine.start(**options)
+
+        self.machine.wait_for_ip_address()
         if not opts.address:
             if opts.trace:
                 print "starting machine %s" % (self.machine.address)
             self.machine.wait_boot()
+
+        # now wait for the other machines to be up
+        for machine in self.machines[1:]:
+            machine.wait_for_ip_address()
+            machine.wait_boot()
```

Comment:
Again, shouldn't we be able to fold wait_for_ip_address() into wait_boot() ... with a fall through in wait_boot() if we already have an ip address.